### PR TITLE
Feature/bcftools

### PR DESCRIPTION
**Pull request by northwestwitch**
Moved and tidied Bcftools.pm. Created tests and modified the calls from mip.pl. Tested several of the subroutines and on rasta and they seem to work.

### CODE REVIEW

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 1]
```diff
@@ -0,0 +1,1203 @@
...
+        },
+        outfile_path    => { strict_type => 1, store => \$outfile_path },
+        infile_path     => { strict_type => 1, store => \$infile_path },
+        stderrfile_path => { strict_type => 1, store => \$stderrfile_path },
+        stderrfile_path_append =>
+          { strict_type => 1, store => \$stderrfile_path_append },
+        FILEHANDLE   => { store       => \$FILEHANDLE },
+        samples_file => { strict_type => 1, store => \$samples_file },
+        constrain => {
+            allow       => [ undef, q{alleles}, q{trio} ],
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: q{alleles}, q{trio} -> qw{ alleles trio }

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 2]
```diff
@@ -0,0 +1,1203 @@
...
+        infile_path     => { strict_type => 1, store => \$infile_path },
+        outfile_path    => { strict_type => 1, store => \$outfile_path },
+        stderrfile_path => { strict_type => 1, store => \$stderrfile_path },
+        stderrfile_path_append =>
+          { strict_type => 1, store => \$stderrfile_path_append },
+        FILEHANDLE  => { store       => \$FILEHANDLE },
+        exclude     => { strict_type => 1, store => \$exclude },
+        soft_filter => { strict_type => 1, store => \$soft_filter },
+        snp_gap => {
+            allow       => qr/^\d+$/,
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: qr/^\d+$/ -> qr/ ^\d+$ /sxm

Comment (henrikstranneheim) [reply lib/MIP/Program/Variantcalling/Bcftools.pm]: throughout

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 3]
```diff
@@ -0,0 +1,1203 @@
...
+
+}
+
+sub bcftools_merge {
+
+    ## bcftools_merge
+
+    ## Function : Perl wrapper for writing bcftools merge recipe to $FILEHANDLE or return commands array. Based on bcftools 1.3.1.
+    ## Returns  : "@commands"
+    ## Arguments: $infile_paths_ref, $outfile_path, $stderrfile_path, $stdoutfile_path, $stderrfile_path_append, $FILEHANDLE, $output_type
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: $stdoutfile_path, $stderrfile_path_append, -> $stderrfile_path_append, $stdoutfile_path,

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 4]
```diff
@@ -0,0 +1,1203 @@
...
+}
+
+sub bcftools_annotate {
+
+    ## bcftools_annotate
+
+    ## Function : Perl wrapper for writing bcftools annotate recipe to $FILEHANDLE or return commands array. Based on bcftools 1.3.1.
+    ## Returns  : "@commands"
+    ## Arguments: $remove_ids_ref, $infile_path, $outfile_path, $stderrfile_path, $stderrfile_path_append, $FILEHANDLE, $samples_file, $headerfile_path, $output_type, $set_id
+    ##          : $remove_ids_ref  => List of annotations to remove
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: Align fat commas

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 5]
```diff
@@ -0,0 +1,1203 @@
...
+        push @commands, q{--samples} . $SPACE . join $COMMA, @{$sample_ids_ref};
+    }
+
+    ## Infile
+    if ($infile_path) {
+
+        push @commands, $infile_path;
+    }
+
+    if ($stderrfile_path) {
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: Remove! Does the test work?

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 6]
```diff
@@ -0,0 +1,1203 @@
...
+
+}
+
+sub bcftools_stats {
+
+    ## bcftools_stats
+
+    ## Function : Perl wrapper for writing bcftools stats recipe to already open $FILEHANDLE or return commands array. Based on bcftools 1.3.1.
+    ## Returns  : "@commands"
+    ## Arguments: $infile_path, $outfile_path, $stderrfile_path, $stdoutfile_path, $stderrfile_path_append, $FILEHANDLE, $append_stderr_info
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: $stdoutfile_path, $stderrfile_path_append, -> change order to reflect rest of header

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 7]
```diff
@@ -0,0 +1,1203 @@
...
+
+    # Stores commands depending on input parameters
+    my @commands = qw{ bcftools stats };
+
+    ## Infile
+    if ($infile_path) {
+
+        push @commands, $infile_path;
+    }
+    ## Outfile
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: Newline

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 8]
```diff
@@ -0,0 +1,1203 @@
...
+    if ($outfile_path) {
+
+        push @commands, q{>} . $SPACE . $outfile_path;
+    }
+
+    if ($stderrfile_path) {
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: Remove!

File: lib/MIP/Program/Variantcalling/Bcftools.pm [hunk 9]
```diff
@@ -0,0 +1,1203 @@
...
+        push @commands, $infile_path;
+    }
+
+## Outfile
+    if ($outfile_path) {
+
+        push @commands, q{>} . $SPACE . $outfile_path;
+    }
+
+    if ($stderrfile_path) {
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Bcftools.pm]: Remove

File: t/bcftools_annotate.t [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+BEGIN {
+
+### Check all internal dependency modules and imports
+## Modules with import
+    my %perl_module;
+
+    $perl_module{q{Script::Utils}} = [qw{ help }];
+
+  PERL_MODULES:
```
Comment (henrikstranneheim) [inline t/bcftools_annotate.t]: Make loop label singular. Throughout

File: t/bcftools_call.t [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        input           => undef,
+        expected_output => $function_base_command,
+    },
+);
+
+## Can be duplicated with %base_argument and/or %specific_argument
+## to enable testing of each individual argument
+my %required_argument = (
+    form_fields_ref => {
+        inputs_ref      => [],
```
Comment (henrikstranneheim) [inline t/bcftools_call.t]: Does this test work. Where does the GQ come from? It does not look like a default in the sub

File: t/bcftools_filter.t [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    exclude => {
+        input           => q{%QUAL<10 || (RPB<0.1 && %QUAL<15)},
+        expected_output => q{--exclude %QUAL<10 || (RPB<0.1 && %QUAL<15)},
+    },
+    soft_filter => {
+        input           => q{LowQual},
+        expected_output => q{--soft-filter LowQual},
+    },
+    snp_gap => {
+        input           => 50,
```
Comment (henrikstranneheim) [inline t/bcftools_filter.t]: Make this into constants i.e. Readonly my $SNP_GAP_FILTER_DISTANCE => 50;

also for indel_gap

File: t/bcftools_norm.t [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+use Modern::Perl qw{ 2014 };
+use warnings qw{ FATAL utf8 };
+use autodie;
+use 5.018;    #Require at least perl 5.18
+use utf8;
+use open qw{ :encoding(UTF-8) :std };
+use charnames qw{ :full :short };
+use Carp;
+use English qw{ -no_match_vars };
+use Params::Check qw{check allow last_error};
```
Comment (henrikstranneheim) [inline t/bcftools_norm.t]: qw{check allow last_error}; -> qw{ check allow last_error };

Comment (henrikstranneheim) [reply t/bcftools_norm.t]: throughout

File: lib/MIP/Program/Variantcalling/Bcftools.pm (hunk 5 above)
Comment (northwestwitch) [reply lib/MIP/Program/Variantcalling/Bcftools.pm]: Of course not, sorry. Yesterday I forgot to push the changes I made after you found this bug!

File: t/bcftools_annotate.t (hunk 10 above)
Comment (northwestwitch) [reply t/bcftools_annotate.t]: Does this apply also to HASHES_OF_ARGUMENTS: at the end of BEGIN ?

Comment (henrikstranneheim) [reply t/bcftools_annotate.t]: ARGUMENT_HASH_REF:

File: t/bcftools_call.t (hunk 11 above)
Comment (northwestwitch) [reply t/bcftools_call.t]: It's the string that is passed to the method in mip.pl. It works, but I'll change inputs_ref into [ [qw{ GQ }] , or something else if you prefer?

**Comment by northwestwitch** [comment]
Ok!

File: t/bcftools_call.t (hunk 11 above)
Comment (henrikstranneheim) [reply t/bcftools_call.t]: Perfect!

File: t/bcftools_filter.t [hunk 14]
```diff
@@ -135,12 +136,12 @@ my %specific_argument = (
...
     snp_gap => {
-        input           => 50,
-        expected_output => q{--SnpGap 50},
+        input           => $SNP_GAP_FILTER_DISTANCE,
+        expected_output => q{--SnpGap} . $SPACE . $SNP_GAP_FILTER_DISTANCE,
     },
     indel_gap => {
-        input           => 100,
-        expected_output => q{--IndelGap 100},
+        input           => $SNP_GAP_FILTER_DISTANCE,
```
Comment (henrikstranneheim) [inline t/bcftools_filter.t]: No, $INDEL_GAP_FILTER_DISTANCE

Comment (northwestwitch) [reply t/bcftools_filter.t]: I know, I was lazy ;)

**Comment by henrikstranneheim** [comment]
🎉 👍

**Comment by northwestwitch** [comment]
YESS :)

**Pull request merged**